core/pkg/server/sender.go

```diff
@@ -441,24 +442,56 @@ func (s *Sender) sendJobFlush() {
 
 	output, err := s.runSummary.CloneTree()
 	if err != nil {
-		s.logger.Error("sender: sendJobFlush: failed to copy run summary", "error", err)
+		s.logger.Error(
+			"sender: sendJobFlush: failed to copy run summary", "error", err,
+		)
 		return
 	}
+	typed_output := data_types.ResolveTypes(output)
 
-	artifact, err := s.jobBuilder.Build(output)
-	if err != nil {
-		s.logger.Error("sender: sendDefer: failed to build job artifact", "error", err)
-		return
-	}
-	if artifact == nil {
-		s.logger.Info("sender: sendDefer: no job artifact to save")
-		return
-	}
-	saver := artifacts.NewArtifactSaver(
-		s.ctx, s.graphqlClient, s.fileTransferManager, artifact, 0, "",
-	)
-	if _, err = saver.Save(s.fwdChan); err != nil {
-		s.logger.Error("sender: sendDefer: failed to save job artifact", "error", err)
+	// If we have a partial job source we just update the
+	// metadata and don't build the job.
+	if s.jobBuilder.PartialJobID != nil {
+		metadata, err := s.jobBuilder.MakeJobMetadata(&typed_output)
+		if err != nil {
+			s.logger.Error(
+				"sender: sendJobFlush: failed to make job metadata",
+				"error",
+				err,
+			)
+			return
+		}
+		_, err = gql.UpdateArtifact(
+			s.ctx, s.graphqlClient, *s.jobBuilder.PartialJobID, &metadata,
+		)
+		if err != nil {
+			s.logger.Error(
+				"sender: sendJobFlush: failed to update partial job metadata",
+				"error",
+				err,
+			)
+			return
+		}
+	} else {
+		artifact, err := s.jobBuilder.Build(output)
+		if err != nil {
+			s.logger.Error(
+				"sender: sendDefer: failed to build job artifact", "error", err,
+			)
+			return
+		}
+		if artifact == nil {
+			s.logger.Info("sender: sendDefer: no job artifact to save")
+			return
+		}
+		saver := artifacts.NewArtifactSaver(
+			s.ctx, s.graphqlClient, s.fileTransferManager, artifact, 0, "",
+		)
+		if _, err = saver.Save(s.fwdChan); err != nil {
+			s.logger.Error(
+				"sender: sendDefer: failed to save job artifact", "error", err,
+			)
+		}
```

Comment:
I'd prefer to move this into  the job builder, that way the intricacies of job building are limited (as much as possible) to the launch package.

Reply:
you think it's better if I pass the sender's api client to `Build` and do the metadata update there and then have it return None?

Reply:
i mean `nil`

Reply:
I think so, my stance is containing the complexities of job building solely within the job building package. However, if the sender's gql client has specific behavior associated with the sender but that doesn't appear to be the case